### ISSUE TITLE
New lint: disallowed_types based on config

### ISSUE THREAD

**Issue by DevinR528** [rust-lang/rust-clippy#7277]
### What it does

Check certain types are not used according to a config list (similar to [disallowed_method](https://rust-lang.github.io/rust-clippy/master/index.html#disallowed_method)).

### Categories

- Nursery

Allow a project to specify types that are not allowed.

### Drawbacks

None :shrug:

### Example

```rust
use std::collections::HashMap;
```

Could be written as:
possibly with a short explanation that is included in the config somehow?
```rust
// Since non of the maps will be large the cost of hashing is greater than the cost of lookup.
use std::collections::BTreeMap;
```

If everyone is ok with this in the nursery I can open a PR.

**Comment by jplatte** [comment]
A great example for where you would disallow `HashMap` (and `HashSet`) specifically is proc-macros, since hash maps introduce non-determinism, which is bad for caching.
